onnxruntime/core/graph/graph_flatbuffers_utils.cc

```diff
@@ -339,4 +340,134 @@ Status LoadAttributeOrtFormat(const fbs::Attribute& fbs_attr,
+    ORT_RETURN_IF(ort_tensor.IsDataTypeString(),
+                  "TensorProto_DataType_STRING is not supported while saving a tensor to ORT format.");
+
+    const auto fbs_tensor_name = builder.CreateString(tensor_name);
+    const auto fbs_tensor_doc_string = builder.CreateString(std::string());
+    const auto fbs_tensor_dims = SaveDims(builder, ort_tensor.Shape().GetDims());
+    flatbuffers::Offset<flatbuffers::Vector<uint8_t>> raw_data = builder.CreateVector(
+        static_cast<const uint8_t*>(ort_tensor.DataRaw()),
+        ort_tensor.SizeInBytes());
+
+    fbs::TensorBuilder tb(builder);
+    tb.add_name(fbs_tensor_name);
+    tb.add_doc_string(fbs_tensor_doc_string);
+    tb.add_dims(fbs_tensor_dims);
+    tb.add_data_type(static_cast<fbs::TensorDataType>(ort_tensor.GetElementType()));
+    tb.add_raw_data(raw_data);
+    fbs_tensor = tb.Finish();
+    return Status::OK();
+  };
+
+  const onnxruntime::Tensor& src_tensor = ort_value.Get<onnxruntime::Tensor>();
+
+  // Check if the tensor is on CPU. If not, we need to copy the tensor to CPU before saving it.
+  if (const auto& tensor_location = src_tensor.Location();
+      tensor_location.device.Type() != OrtDevice::CPU) {
+    InlinedVector<uint8_t> tensor_data_buffer{};
+    tensor_data_buffer.resize(src_tensor.SizeInBytes());
+    const OrtMemoryInfo cpu_alloc_info{onnxruntime::CPU, OrtDeviceAllocator};
+    onnxruntime::Tensor dst_tensor{src_tensor.DataType(), src_tensor.Shape(), tensor_data_buffer.data(), cpu_alloc_info};
+    ORT_RETURN_IF_ERROR(copy_tensor(src_tensor, dst_tensor));
+    ORT_RETURN_IF_ERROR(ort_tensor_to_fbs_tensor(tensor_name, dst_tensor, fbs_tensor));
+  } else {
+    ORT_RETURN_IF_ERROR(ort_tensor_to_fbs_tensor(tensor_name, src_tensor, fbs_tensor));
+  }
+
+  return Status::OK();
+}
+
+#define UNPACK_TESOR_WITH_TYPE(T)                                                                          \
+  ORT_RETURN_IF_ERROR(onnxruntime::utils::UnpackTensor(unused_tensor_proto, fbs_tensor.raw_data()->Data(), \
+                                                       fbs_tensor.raw_data()->size(),                      \
+                                                       dst_tensor->MutableData<T>(),                       \
+                                                       static_cast<size_t>(dst_tensor->Shape().Size())));  \
+  break;
+
+Status LoadOrtValueOrtFormat(const fbs::Tensor& fbs_tensor,
+                             std::string& tensor_name, OrtValue& ort_value) {
+  // The assumption is that the flatbuffer buffer will be destructed once the checkpoint has been loaded.
+  // And so, we must allocate a buffer where the tensor data can be copied using the cpu allocator.
+  // This buffer is owned by the OrtValue.
+  static const CPUExecutionProviderInfo info;
+  static const CPUExecutionProvider cpu_provider(info);
+  static const AllocatorPtr cpu_allocator = cpu_provider.GetAllocator(OrtMemTypeDefault);
```

Comment:
instead of these static variables (including a AllocatorPtr, which is a shared_ptr), should we pass in an allocator?